tools/generate-rust-dashboards/src/schema.rs

```diff
@@ -0,0 +1,706 @@
+        }
+    }
+}
+
+impl Target {
+    pub fn timeseries(sql: impl Into<String>) -> Self {
+        Self {
+            format: TargetFormat::Timeseries,
+            raw_query: true,
+            raw_sql: sql.into(),
+        }
+    }
+
+    pub fn table(sql: impl Into<String>) -> Self {
+        Self {
+            format: TargetFormat::Table,
+            raw_query: true,
+            raw_sql: sql.into(),
+        }
+    }
+}
+
+pub struct DashboardBuilder {
+    pub dashboard: Dashboard,
+    col: u32,
+    row: u32,
+    current_row_height: u32,
+}
+
+impl DashboardBuilder {
+    pub fn new(title: impl Into<String>, uid: impl Into<String>) -> Self {
+        Self {
+            dashboard: Dashboard {
+                title: title.into(),
+                uid: uid.into(),
+                ..Dashboard::default()
+            },
+            col: 0,
+            row: 0,
+            current_row_height: 0,
+        }
+    }
+
+    pub fn add_variable(&mut self, v: impl Into<Variable>) {
+        self.dashboard.templating.list.push(v.into());
+    }
+
+    /// Add an `application` variable that the user can select
+    pub fn add_application_variable(&mut self, config: &TeamConfig) -> Result<()> {
+        let applications = config.applications();
+
+        let first_application = applications
+            .iter()
+            .next()
+            .ok_or_else(|| anyhow!("Application list empty for {}", config.team_name))?;
+
+        self.add_variable(CustomVariable {
+            label: "Application".into(),
+            name: "application".into(),
+            query: applications
+                .iter()
+                .map(|a| format!("{a} : {}", a.slug()))
+                .collect::<Vec<_>>()
+                .join(","),
+            current: CustomVariableSelection::single(first_application.slug()),
+            ..CustomVariable::default()
+        });
+        Ok(())
+    }
+
+    /// Add an `channel` variable that the user can select a release channel from
+    pub fn add_channel_variable(&mut self) {
+        self.add_variable(CustomVariable {
+            label: "Channel".into(),
+            name: "channel".into(),
+            multi: false,
+            query: "nightly,beta,release".into(),
+            ..CustomVariable::default()
+        });
+    }
+
+    // Add a `filter_sql` variable
+    //
+    // This is a WHERE condition to filter queries on, based on the LogPanel's `Filters` variable.
+    //
+    // This converts the "ad-hoc filter" syntax to SQL.  It's pretty gross, but mostly works.
+    pub fn add_filter_sql_variable(&mut self) {
+        self.add_variable(
+            QueryVariable {
+                name: "filter_sql".into(),
+                hide: VariableHide::Variable,
+                datasource: Datasource::bigquery(),
+                query: QueryVariableQuery::from_sql(
+                    r#"SELECT IF(STRPOS('${Filters}', '=') <> 0, REPLACE(REPLACE('${Filters}', '",', '" AND '), '\n', '\\n'), 'true')"#,
```

Comment:
If i'm understanding correctly this might make some of the queries vulnerable here where it's used in `{filter_sql}` directly into the SQL statement (injection)? Idk how Grafana handles some of these but maybe more of a warning or something here.